frontend/integration-tests/tests/kubevirt/template.wizard.scenario.ts

```diff
@@ -80,20 +81,15 @@ describe('Kubevirt create VM template using wizard', () => {
     const vm = new VirtualMachine(vmConfig);
 
     it(`Create VM template using ${configName}.`, async() => {
-      addLeakableResource(leakedResources, template.asResource());
       await template.create(templateConfig);
-
-      // Verify the template can be used to create VM
-      addLeakableResource(leakedResources, vm.asResource());
-      await vm.create(vmConfig);
-
-      // Remove VM
-      deleteResource(vm.asResource());
-      removeLeakableResource(leakedResources, vm.asResource());
-
-      // Remove template
-      deleteResource(template.asResource());
-      removeLeakableResource(leakedResources, template.asResource());
+      await withResource(leakedResources, template.asResource(), async() => {
+        // Verify template can be used to create a VM
+        await vm.create(vmConfig);
```

Comment:
shouldn't the vm be added to leakedResources? What about creating a second method `withResources(leakedResources, resources, callback)` ?

Reply:
@suomiy I actually thought about this.
Thing is, with current implementation, if `create` resolves without an exception, it means that the VM must have been created and in this case it is then save to just delete it. 
The problem is when it fails, because the configuration may have `startOnCreation: true`, in which case the `create` method waits for `Running`. And that can potentially time out too. But at this point the VM is already created. So if this method throws an exception, there is no way to tell whether it has or has not actually created anything.

I have the following idea:
the `create` methods need to catch exceptions and return created object or i.e. `undefined`
Then the `withResource` could be updated to:
```
export async function withResource(resourceSet: Set<string>, resource: any, callback: Function, keepResource: boolean = false) {
  if (resource != undefined) {
    addLeakableResource(resourceSet, resource);
    await callback();
    if (keepResource !== true) {
      deleteResource(resource);
    }
    removeLeakableResource(resourceSet, resource);
  } else {
    throw new Error('Failed to create given resource')
}
```
and in our code we could use it as:
```
await withResource(leakedResources, vm.asResource(), async(vm: VirtualMachine) = {
  ...
});
```
The only problem remaining is with remove/addLeakableResource methods, which currently take `{kind: string, metadata: {name, namespace}}`. I guess we could check for types and distinguish between passing manifest object and an object derived from `DetailView` class.
Suggestions?

It's quite a big change thought, I'd prefer merging this with addressing the previous comments and leave this for another PR so that our other PRs are not blocked by this. What do you think?

Reply:
> It's quite a big change thought, I'd prefer merging this with addressing the previous comments and leave this for another PR so that our other PRs are not blocked by this. What do you think?

Sure we can leave it for later. But my main problem here is still that we are not disposing of the created vm (which was being done before this change). Please correct me if I am wrong.


> I have the following idea:
the create methods need to catch exceptions and return created object or i.e. undefined
Then the withResource could be updated 

I would prefer using exceptions. You could create a custom Error to distinguish between these cases (creation vs  starting failed)

Reply:
> Sure we can leave it for later. But my main problem here is still that we are not disposing of the created vm (which was being done before this change). Please correct me if I am wrong.
Right, thanks. Fixed.

> I would prefer using exceptions. You could create a custom Error to distinguish between these cases (creation vs starting failed)
Yes, sounds reasonable.